plugin/call_hierarchy.py

```diff
@@ -0,0 +1,168 @@
+            tooltip="{}:{}".format(self._simple_path(element['uri']), element['selectionRange']['start']['line'] + 1),
+            command_url=command_url
+        )
+
+    def _simple_path(self, uri: DocumentUri) -> str:
+        scheme, path = parse_uri(uri)
+        session = self.weaksession()
+        if not session or scheme != 'file':
+            return path
+        simple_path = simple_project_path([Path(folder.path) for folder in session.get_workspace_folders()], Path(path))
+        return str(simple_path) if simple_path else path
+
+    def _handle_incoming_calls_async(
+        self, response: Optional[List[CallHierarchyIncomingCall]]
+    ) -> List[CallHierarchyItem]:
+        return [incoming_call['from'] for incoming_call in response] if response else []
+
+    def _handle_outgoing_calls_async(
+        self, response: Optional[List[CallHierarchyOutgoingCall]]
+    ) -> List[CallHierarchyItem]:
+        return [outgoing_call['to'] for outgoing_call in response] if response else []
+
+
+class LspCallHierarchyCommand(LspTextCommand):
+
+    capability = 'callHierarchyProvider'
+
+    def is_visible(self, event: Optional[dict] = None, point: Optional[int] = None) -> bool:
+        if self.applies_to_context_menu(event):
+            return self.is_enabled(event, point)
+        return True
+
+    def run(self, edit: sublime.Edit, event: Optional[dict] = None, point: Optional[int] = None) -> None:
+        self._window = self.view.window()
+        session = self.best_session(self.capability)
+        if not session:
+            return
+        position = get_position(self.view, event, point)
+        if position is None:
+            return
+        params = cast(CallHierarchyPrepareParams, text_document_position_params(self.view, position))
+        request = Request.prepareCallHierarchy(params, self.view)
+        session.send_request_async(request, partial(self._handle_response_async, weakref.ref(session)))
+
+    def _handle_response_async(
+        self, weaksession: 'weakref.ref[Session]', response: Optional[List[CallHierarchyItem]]
+    ) -> None:
+        if not self._window or not self._window.is_valid():
+            return
+        if not response:
+            self._window.status_message("Call hierarchy not available")
+            return
+        session = weaksession()
+        if not session:
+            return
+        data_provider = CallHierarchyDataProvider(weaksession, CallHierarchyDirection.IncomingCalls, response)
+        header = 'Call Hierarchy: Callers of… {}'.format(
+            make_command_link('lsp_call_hierarchy_toggle', "⇄", {
+                'session_name': session.config.name,
+                'direction': CallHierarchyDirection.OutgoingCalls,
+                'root_elements': response
+            }, tooltip="Show outgoing calls"))
+        new_tree_view_sheet(self._window, "Call Hierarchy", data_provider, header, flags=sublime.ADD_TO_SELECTION)
```

Comment:
A question popped into my head...

Would it make sense to actually open the call hierarchy view in its own, primary tab instead of attaching it to the active tab and then ending up with three tabs after clicking something? It takes too much space to be usable with three tabs opened at once IMO.

Reply:
Certainly possible, but since the call hierarchy doesn't use much horizontal space (unless you have excessively long function names...), I thought it would be nice to have it as a side-by-side view, so that you don't lose the context. And you can just click on the call hierarchy tab in the tab bar if you want to exit the side-by-side mode and only focus this tab.

I have made quite good experiences with a custom "Show Documentation" command from LSP-julia, which I also open in the same way as a side-by-side HtmlSheet. But of course it's very subjective and I haven't really used the call hierarchy functionality in practice yet.

We could also consider to add a `side_by_side` command argument, like the "Goto..." commands already have. It probably wouldn't be very useful for the context menu, because the menu can't easily be configured by users. But it might be useful in case someone wants to add a custom key binding for it.

Reply:
> I thought it would be nice to have it as a side-by-side view, so that you don't lose the context.

Just gonna throw another idea out there but it might hacky or hard to do

Open call hierarchy in a new new tab and then add the previously open file to selection (I guess that would technically open it again in the same location). Then allow the user to just click call hierarchy items to replace the view on the right.

> I have made quite good experiences with a custom "Show Documentation" command from LSP-julia

I haven't tried it but I imagine documentation sheet doesn't need to open further sheets on the right which is the main thing here?

> We could also consider to add a `side_by_side` command argument, like the "Goto..." commands already have.

I suppose the default behavior could be controlled by an LSP option (as much as we hate adding new options).

Reply:
Also to throw another idea. (not for now) 

Instead of a sheet for the call hierarchy use a regular View.
Benefits, it allows to have a cursor.
- We could than move in the View with the cursor, 
and add a keybinding for example. `space` or `enter` to toggle the nodes. That way we don't need to use the mouse.
- We can lock the view to not be editable.

Reply:
I agree that the sheet is not very Sublime-like since it's not keyboard accessible. Plain view should allow more or less equivalent functionality.

But I'm not suggesting to change it now and also it wouldn't really affect this specific issue I've mentioned in this thread.

Reply:
I don't really think that a regular view would be better:

* It probably wouldn't look as nice with the mouse cursor always visible and everthing forced as monospace font (including the header) - the HtmlSheet allows more styling
* It would be unexpected if you bind the left and right arrow keys to collapse / expand a tree item like in the sidebar, because there is still a cursor visible which you couldn't move to the left/right then.
* If the disclosure buttons (triangles) should be clickable with the mouse, then they must be phantoms, which probably would make things very ugly from the implementation side, because you need to keep track of all the positions when you update the icons, while new rows for child items are being added or removed. Because phantom's contents can't be updated, so you would need to recreate them each time when an icon should change.
* You can't have the kinds rendered like an icon, unless they are phantoms.
* You can't click the function name to open the location, unless it is a phantom too. But then basically everything in the view is a phantom again... so you could use the HtmlSheet in the first place instead :-)
* Can't have a tooltip when hovering over the function names if they are plain text.
* If you want to style the colors, then you need to implement a syntax definition for this view.

So basically it amounts to only keyboard navigation (regular view) vs. only mouse navigation (HtmlSheet), and I would say for such trees mouse navigation feels more natural.

The only other advantage I see, is that you could place a regular view as an output panel at the bottom. But I think that's also not really beneficial for the call hierarchy tree view, because it usually needs more vertical space than horizontal space.

If ST some day adds a sidebar API to render custom trees, then yes, we should switch to that, but right now the HtmlSheet seems to be the simplest and best we can do from my point of view.

---

> Just gonna throw another idea out there but it might hacky or hard to do
> 
> Open call hierarchy in a new new tab and then add the previously open file to selection (I guess that would technically open it again in the same location). Then allow the user to just click call hierarchy items to replace the view on the right.

I just tried, and `window.open_file("...", flags=sublime.ADD_TO_SELECTION|sublime.SEMI_TRANSIENT)` for the original view would open it a second time to the right. But in order to replace it with new views, it needs to be semi-transient, and as already mentioned the SEMI_TRANSIENT flag doesn't work reliably in the API functions currently. Also I haven't/can't check what happens if it has unsaved changes and you want to replace it with another view when clicking on the call hierarchy items. It would be annyoing if ST asks to save the file in that case.

Reply:
I can merge it and we can refine it later or we can wait until next ST version and make use of SEMI_TRANSIENT then.

I'm not too happy with how it works now (by default) where if the clicked entry is in the same file it will scroll to it but if it's in another then it will open new file. I would like this to be more consistent.

What would you prefer? Get this in or wait?

Reply:
> What would you prefer? Get this in or wait?

Based on the previous release frequency, the next stable release might be several months or even longer away from now. We probably don't want to wait *that* long.

I was thinking if it would be better to make the behavior when holding <kbd>Ctrl</kbd> / <kbd>Cmd</kbd> while clicking on an item, to be the default, even if no modifier key is used. Because currently there is no indication that you are able to use the modifier keys when clicking the item, so I could imagine that most people don't find out about this ability at all. But on the other hand that would be inconsistent with how ST's built-in features work, for example the links from the "Definitions popup" (this one shows a hint for the modifier key in the tooltip, but in our case I already used the tooltip for the file location of the link). And in general I'm in favor of consistency and to follow ST default behavior.

Another option would be to add an `if int(sublime.version()) > 4144` check (I think this bug is `https://github.com/sublimehq/sublime_text/issues/5729` and was fixed in that version) and have two different code paths, at least until the next stable ST release. But I'm not sure if we want to add this additional complexity and maintenance effort.

Reply:
> I was thinking if it would be better to make the behavior when holding Ctrl / Cmd while clicking on an item, to be the default

I think that would be better than the current behavior.

> Because currently there is no indication that you are able to use the modifier keys when clicking the item,

Since we have a lot of flexibility with minihtml, we can easily add a text that would mention that in the output.

> have two different code paths, at least until the next stable ST release

I would be fine with that if it's just temporary and wouldn't require a ton of extra code.

Reply:
I added a commit to always open the links in side-by-side mode now. I think I've tried this already before https://github.com/sublimelsp/LSP/pull/2151#discussion_r1058283725, but the trick was to also use the `sublime.SEMI_TRANSIENT` flag to make it work.

Also it will put the HtmlSheet on the left side now, and already open the location of the root item to the right.

* Is it better than before?
* Does it work on the latest dev build?

If not, I could revert the last commit.

Reply:
It works really nice now besides the fact that we are hitting https://github.com/sublimehq/sublime_text/issues/2411 for the initially open file. Switching to another file and back makes it work for the initial file.


https://user-images.githubusercontent.com/153197/213924885-5702dee5-730d-4519-8f73-57bf64ad13dc.mov

Reply:
I've pushed a work around for that. I think it's good to have it as it should help non-call-hierarchy cases also.

Reply:
I'll put this change in a separate PR (https://github.com/sublimelsp/LSP/pull/2174) so that we have it on record and change log. Otherwise this looks good to me.

Reply:
> Is it better than before?
Yes, definitely!  
Pretty happy with how this works